pages/object/Vibes.tsx

```diff
@@ -0,0 +1,125 @@
+
+interface IProps {
+  tokenId: string | string[]
+}
+
+// VIBES ticker refresh in ms
+const REFRESH_VIBES_INTERVAL = 100;
+
+const SICK_VIBES_SITE_URL = 'https://www.sickvibes.xyz';
+
+// VIBES after render
+interface VibesAtRender {
+  initialClaimableVibes: BigNumber
+  initialDateTime: Date
+}
+
+const formatVibes = (vibes: BigNumber, decimal = 18, toFixed = 3): string => {
+  // convert to base 10
+  const decimalVibes = vibes.toString();
+  
+  // Split into two parts
+  const indexForDecimalPoint = decimalVibes.length - decimal;
+  const wholeNumber = decimalVibes.slice(0, indexForDecimalPoint);
+  const decimalNumber = decimalVibes.slice(indexForDecimalPoint);
+
+  // Format and concatenate
+  const formattedWholeNumber = wholeNumber.replace(/\B(?=(\d{3})+(?!\d))/g, ",");
+  const formattedDecimalNumber = decimalNumber.slice(0, toFixed);
+  const formattedVibes = formattedWholeNumber + '.' + formattedDecimalNumber;
+
+  // Return dat pretty boi
+  return formattedVibes;
+};
+
+/**
+ * Calculate live claimable VIBES and return in human-readable form
+ */
+const calculateLiveInfusedVibes = (initialVibes: VibesAtRender, dailyRate: BigNumber) => {
+  const now = new Date();
+  const { initialClaimableVibes, initialDateTime } = initialVibes;
+  const msElapsed = now.getTime() - initialDateTime.getTime();
+
+  // do math
+  const vibesToAdd = dailyRate.mul(msElapsed).div(24 * 60 * 60 * 1000);
+  const currentClaimableVibes = initialClaimableVibes.add(vibesToAdd);
+
+  // format it real cute
+  const formattedVibes = formatVibes(currentClaimableVibes);
+  return formattedVibes;
+};
+
+const Vibes = ({ tokenId }: IProps) => {
+  const { account } = useWeb3React<Web3Provider>()
+  const [tokenInfo, setTokenInfo] = useState<Record<string, unknown> | undefined>();
+  const [claimableVibes, setClaimableVibes] = useState<string>('');
+  const [startDateTime, _] = useState<Date>(new Date());
+
+  const vibesTokenUrl = `${SICK_VIBES_SITE_URL}/tokens/${process.env.NEXT_PUBLIC_CONTRACT_ID}/${tokenId}`;
+
+  async function getVibes() {
+    const contract = new ethers.Contract(
+      process.env.NEXT_PUBLIC_VIBES_CONTRACT_ID,
+      VIBES_WELLSPRING_ABI,
+      getNetworkLibrary(),
+    )
+    const token = await contract.getToken(process.env.NEXT_PUBLIC_CONTRACT_ID, tokenId);
+    if (token && token.isSeeded) {
+      setTokenInfo(token);
+    } 
+  }
+
+  function getClaimableVibes() {
+    if (tokenInfo) {
+      const claimableVibes = calculateLiveInfusedVibes(
+        {
+          initialClaimableVibes: tokenInfo.claimable as BigNumber,
+          initialDateTime: startDateTime 
+        },
+        tokenInfo.dailyRate as BigNumber,
+      );
+      setClaimableVibes(claimableVibes);
+    }
+  }
+
+  useEffect(() => {
+    getVibes();
+  }, [account])
+
+  useEffect(() => {
+    if (tokenInfo) {
+      const h = setInterval(getClaimableVibes, REFRESH_VIBES_INTERVAL);
+      return () => clearInterval(h);
+    }
+  })
```

Comment:
read that useEffect executes in order; check me on this fam

Intention here is that getting claimable vibes should only be executed after token info is loaded. Is this the way? I checked that it worked, but is this the best way to achieve that effect?

Reply:
Yep, 100% correct that `useEffect` blocks execute in order, and re: your intention, this is exactly how you would do that, i.e. some state change (`tokenInfo`) needs to trigger some side effect (`setInterval`).

Prob not a huge deal since you are wrapping this code in a conditional, but I would also add a dependency array with `[tokenInfo]` here, which basically instructs the `useEffect` to only execute if `tokenInfo` has changed since last render. The way it's currently written, it will execute on every render.

Like I said, def not a huge deal as it stands, but if we were to add any state change code (or anything else that triggers a re-render) to this effect block in the future, it would create an infinite render loop (state updates which triggers effect to fire, which updates state, which triggers effect to fire, etc.)

Reply:
oh i likeeee that 🔥 gonna update